drivers/adc/adm1177/iio_adm1177.c

```diff
@@ -0,0 +1,251 @@
+* IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+* LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************************************************************/
+#include "iio_adm1177.h"
+#include "adm1177.h"
+#include "no_os_error.h"
+#include "no_os_util.h"
+#include "no_os_alloc.h"
+#include "no_os_delay.h"
+
+static int adm1177_iio_read_raw(void *device, char *buf, uint32_t len,
+				const struct iio_ch_info *channel,
+				intptr_t priv)
+{
+	struct adm1177_iio_dev *iiodev = (struct adm1177_iio_dev *)device;
+	struct adm1177_dev *desc = (struct adm1177_dev *)iiodev->adm1177_dev;
+	int32_t ret, value;
+	uint16_t conv_voltage;
+	uint16_t conv_current;
+
+	switch(channel->type) {
+	case IIO_VOLTAGE:
+		ret = adm1177_read(desc, &conv_voltage, &conv_current);
+		if(ret)
+			return ret;
+
+		value = no_os_sign_extend32((uint32_t)conv_voltage, 11);
+
+		break;
+	case IIO_CURRENT:
+		ret = adm1177_read(desc, &conv_voltage, &conv_current);
+		if(ret)
+			return ret;
+
+		value = no_os_sign_extend32((uint32_t)conv_current, 11);
+
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return iio_format_value(buf, len, IIO_VAL_INT, 1, &value);
+}
+
+static int adm1177_iio_read_scale(void *device, char *buf, uint32_t len,
+				  const struct iio_ch_info *channel,
+				  intptr_t priv)
+{
+	int32_t vals[2];
+	struct adm1177_iio_dev *iio_adm1177;
+	struct adm1177_dev *adm1177;
+
+	if(!device)
+		return -EINVAL;
+
+	iio_adm1177 = device;
+
+	if(!iio_adm1177->adm1177_dev)
+		return -EINVAL;
+
+	adm1177 = iio_adm1177->adm1177_dev;
+
+	switch(channel->type) {
+	case IIO_VOLTAGE:
+		if(adm1177->vrange == ADM1177_VRANGE_LOW)
+			vals[0] = ADM1177_V_FULLSCALE_0;
+		else {
+			vals[1] = ADM1177_V_FULLSCALE_1;
+		}
+		vals[1] = 12;
+		return iio_format_value(buf, len, IIO_VAL_FRACTIONAL_LOG2, 2,
+					vals);
+	case IIO_CURRENT:
+		vals[0] = 105840;
+		vals[1] = 4096 * ADM1177_R_SENSE;
+		return iio_format_value(buf, len, IIO_VAL_FRACTIONAL, 2,
+					vals);
+	default:
+		return -EINVAL;
+	}
+}
+
+static int adm1177_iio_read_samples(void *device, int *buff, uint32_t samples)
+{
+	uint16_t conv_voltage;
+	uint16_t conv_current;
+	uint32_t raw_voltage;
+	uint32_t raw_current;
+	struct adm1177_iio_dev *iio_adm1177;
+	struct adm1177_dev *adm1177;
+
+	if(!device)
+		return -EINVAL;
+
+	iio_adm1177 = (struct adm1177_iio_dev *)device;
+
+	if(!iio_adm1177->adm1177_dev)
+		return -EINVAL;
+
+	adm1177 = iio_adm1177->adm1177_dev;
+
+	for(uint32_t i = 0; i < samples*2; i++) {
```

Comment:
In this loop you need to check if the channel is enabled and populate samples only if it's enabled:
https://github.com/analogdevicesinc/no-OS/blob/master/drivers/accel/adxl355/iio_adxl355.c#L933

The driver underneath and the chip, they will work with the assumption of V_CONT + I_CONT all the time, but IIO wants to be able to enable/disable channels so you'll have to do that.